File: examples/node/scan.js
// A runnable Node.js example: scan a small universe through the binding.
//
//   ( cd bindings/node && npm install && npm run build )
//   ( cd examples/node && npm install && node scan.js )

"use strict";

const { Screener, version } = require("wickra-screener");

const SPEC = JSON.stringify({
  universe: ["AAA", "BBB"],
  condition: {
    type: "cmp",
    left: { kind: "price", field: "close" },
    op: "gt",
    right: { kind: "const", value: 10.0 },
  },
});

const candle = (close) => ({
  time: 1,
  open: close,
  high: close,
  low: close,
  close,
  volume: 1.0,
});

const screener = new Screener(SPEC);
const response = screener.command(
  JSON.stringify({
    cmd: "scan",
    data: { AAA: [candle(5.0)], BBB: [candle(15.0)] },
  }),
);
const report = JSON.parse(response);

console.log("wickra-screener", version());
console.log(response);
for (const match of report.matches) {
  console.log(`  matched: ${match.symbol}`);
}

File: examples/python/scan.py
"""A runnable Python example: scan a small universe through the binding.

    pip install wickra-screener
    python examples/python/scan.py
"""

import json

from wickra_screener import Screener

SPEC = json.dumps(
    {
        "universe": ["AAA", "BBB"],
        "condition": {
            "type": "cmp",
            "left": {"kind": "price", "field": "close"},
            "op": "gt",
            "right": {"kind": "const", "value": 10.0},
        },
    }
)


def candle(close: float) -> dict:
    return {
        "time": 1,
        "open": close,
        "high": close,
        "low": close,
        "close": close,
        "volume": 1.0,
    }


def main() -> None:
    screener = Screener(SPEC)
    response = screener.command(
        json.dumps(
            {
                "cmd": "scan",
                "data": {"AAA": [candle(5.0)], "BBB": [candle(15.0)]},
            }
        )
    )
    report = json.loads(response)

    print(f"wickra-screener {Screener.version()}")
    print(response)
    for match in report["matches"]:
        print(f"  matched: {match['symbol']}")


if __name__ == "__main__":
    main()
